### ISSUE TITLE
Write about lower quality inference in Flow

### ISSUE THREAD

**Issue by niieani** [niieani/typescript-vs-flowtype#9]
```ts
function test(param : Array<number | string>) {
  return param
}

const compatibileVar : Array<number> = [1, 2, 3]

test(compatibileVar)
```

This [works properly in TypeScript](http://www.typescriptlang.org/play/#src=function%20test(param%20%3A%20Array%3Cnumber%20%7C%20string%3E)%20%7B%0D%0A%20%20return%20param%0D%0A%7D%0D%0A%0D%0Aconst%20compatibileVar%20%3A%20Array%3Cnumber%3E%20%3D%20%5B1%2C%202%2C%203%5D%0D%0A%0D%0Atest(compatibileVar)%0D%0A%0D%0A), but [fails in Flow](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVUCuA7AxgFwEs5sx8BTAZ3wAoAHAQwCcGBbMALjAEEmWBPADzZMrAEbkmYAD5hqTQtgDmAPgCUYAN6owYJuXyYmpRi1aoAvulwlqYG60ZExhGOQBqzTjz4MhI8UkVMABeMABtAEYAGjAAJliAZgBddApqGgcnQhc3TyY1dCA) with the error:

```ts
3: function test(param : Array<number | string>) {
                                        ^ string. This type is incompatible with
7: const compatibileVar : Array<number> = [1, 2, 3]
                                ^ number
```

**Comment by vkurchatkin** [comment]
Just to clarify: Flow is being more strict here, not TS

**Comment by niieani** [comment]
@vkurchatkin I know it's being more strict, I just can't think of a reason why would you want this behavior. Can you give a reason as to why this is better?

**Comment by vkurchatkin** [comment]
This is better because it's safe. Typescript's behavior is unsafe:

```ts
function test(param: Array<number | string>) {
  param.push('foo')
}

const compatibleVar : Array<number> = []

compatibleVar[0].toFixed(); // Runtime error
```

**Comment by niieani** [comment]
@vkurchatkin ok, this makes sense, but only if you're mutating data. It would be nice to be able to opt-in or opt-out of this, since with modern programming / immutability / functional dogma one really shouldn't be doing what `test` in the above example does (i.e. mutating the parameters). 

I also think what you gave as an example it's an edge case, so personally, I believe TypeScript's trade-off for a bit of unsoundness goes a long way in making it easier to develop in general.

(btw. you missed one line: `test(compatibleVar)`, but I inferred what you were trying to say :))

**Comment by vkurchatkin** [comment]
Well, in Flow you can just do:

```js
function test(param: $ReadOnlyArray<number | string>) {
}
```

And it's going to work

**Comment by niieani** [comment]
That's great to know, however this is an undocumented feature. :(

**Comment by leoasis** [comment]
Maybe this blog post can be helpful regarding this particular Flow behavior? https://medium.com/@forbeslindesay/covariance-and-contravariance-c3b43d805611#.6oemuj9gn

Unless you're already familiar with it, in any case maybe it helps for posterity.

**Comment by niieani** [comment]
@leoasis I was aware of variance (it's documented), it's only the `$ReadOnlyArray` that I didn't know about :) But thanks, the article is pretty useful!

**Comment by RyanCavanaugh** [comment]
Probably good to talk about Flow's treatment of mutated arrays in the same section:
```js
const arr = ['1', 0];
arr.sort();
// Error in TS, not in Flow, throws at runtime
arr[0].substr(0); 
```

**Comment by vkurchatkin** [comment]
Interesting example, I've already filed this issue here: https://github.com/facebook/flow/issues/3735

The thing is, Typescript's behaviour seems to be not that different. In your example TS only catches the error because it doesn't try to infer precise type of `arr`. But if you add an annotation it's going to allow the same behaviour:

```js
const arr: [string, number] = ['1', 0];
arr.sort();
arr[0].substr(0)
```

Flow, on the other hand, will show an error in this case.

**Comment by RyanCavanaugh** [comment]
> Flow, on the other hand, will show an error in this case.

Only because tuples are read-only. The equivalent TS code would be:
```ts
const arr: ReadonlyArray<number | string> = ['1', 0];
arr.sort();
```
or
```ts
const arr: { 0: string, 1: number } = ['1', 0];
arr.sort();
```
which both show the same error.

Probably TS tuples should inherit from some other Array type that allows mutation but doesn't allow reordering operations. Allowing `sort` / `splice` / etc on `[number, string]` is certainly wrong.

**Comment by vkurchatkin** [comment]
> Only because tuples are read-only. The equivalent TS code would be:

No, they are not really read-only, they just have methods of `$ReadOnlyArray`.

> Probably TS tuples should inherit from some other Array type that allows mutation but doesn't allow reordering operations. Allowing sort / splice / etc on [number, string] is certainly wrong.

That's basically what Flow does
